File: src/main/java/ir/alireza/sadeghi/shiraz/AI/weight/optimisation/TrainingInstances.java
package ir.alireza.sadeghi.shiraz.ai.weight.optimisation;

import ir.alireza.sadeghi.shiraz.ai.ReadMatrix;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public class TrainingInstances {

    private final static Logger logger = LogManager.getLogger(TrainingInstances.class);
    static double[][] AggressiveInstance = ReadMatrix.readIn(System.getProperty("user.dir") + ReadMatrix.slash + "\\src\\main\\java\\ir\\alireza\\sadeghi\\shiraz\\ai\\StartingAI\\Aggressive.txt");
    static double[][] NeutralInstance = ReadMatrix.readIn(System.getProperty("user.dir") +  ReadMatrix.slash + "\\src\\main\\java\\ir\\alireza\\sadeghi\\shiraz\\ai\\StartingAI\\Neutral.txt");
    static double[][] DefensiveInstance = ReadMatrix.readIn(System.getProperty("user.dir") +  ReadMatrix.slash + "\\src\\main\\java\\ir\\alireza\\sadeghi\\shiraz\\ai\\StartingAI\\Defensive.txt");

    public static void randInstanceCreation() {

        ReadMatrix.fileNumber = 0;

        double[][] RandNeutralInstance = NeutralInstance;
        double[][] RandAggressiveInstance = AggressiveInstance;
        double[][] RandDefensiveInstance = DefensiveInstance;
        double randomInterval = 0.1;

        for (int i = 0; i < RandNeutralInstance.length; i++) {
            for (int j = 0; j < RandNeutralInstance[0].length; j++) {

                RandNeutralInstance[i][j] = NeutralInstance[i][j] - randomInterval + Math.random() * 2 * randomInterval;       // midpoint of [a,b] is value of NeutralInstance[i][j]

                if (RandNeutralInstance[i][j] < 0) {
                    RandNeutralInstance[i][j] = 0;
                }
            }
        }

        for (int i = 0; i < RandAggressiveInstance.length; i++) {
            for (int j = 0; j < RandAggressiveInstance[0].length; j++) {

                RandAggressiveInstance[i][j] = AggressiveInstance[i][j] - randomInterval + Math.random() * 2 * randomInterval;       // midpoint of [a,b] is value of AggressiveInstance[i][j]

                if (RandAggressiveInstance[i][j] < 0) {
                    RandAggressiveInstance[i][j] = 0;
                }
            }
        }

        for (int i = 0; i < RandDefensiveInstance.length; i++) {
            for (int j = 0; j < RandDefensiveInstance[0].length; j++) {

                RandDefensiveInstance[i][j] = DefensiveInstance[i][j] - randomInterval + Math.random() * 2 * randomInterval;       // midpoint of [a,b] is value of DefensiveInstance[i][j]

                if (RandDefensiveInstance[i][j] < 0) {
                    RandDefensiveInstance[i][j] = 0;
                }
            }
        }

        normalize(RandNeutralInstance);
        normalize(RandAggressiveInstance);
        normalize(RandDefensiveInstance);

        ReadMatrix.readOut(RandNeutralInstance,"Trainers");
        ReadMatrix.readOut(RandAggressiveInstance,"Trainers");
        ReadMatrix.readOut(RandDefensiveInstance,"Trainers");
    }

    //Normalisation:
    private static void normalize(double[][] inputMatrix) {

        for (int l = 0; l < inputMatrix.length; l++) {
            double sum = 0;
            for (int m = 0; m < inputMatrix[0].length; m++) {
                sum += inputMatrix[l][m];
            }
            for (int n = 0; n < inputMatrix.length; n++) {
                inputMatrix[l][n] = (inputMatrix[l][n]) / sum;
            }
        }
    }
}

File: src/main/java/ir/alireza/sadeghi/shiraz/AI/weight/optimisation/TrainingInstances.kt
package ir.alireza.sadeghi.shiraz.ai.weight.optimisation

import ir.alireza.sadeghi.shiraz.ai.ReadMatrix
import org.apache.logging.log4j.LogManager

object TrainingInstances {
    private val logger = LogManager.getLogger(
        TrainingInstances::class.java
    )
    private var aggressiveInstance =
        ReadMatrix.readIn(System.getProperty("user.dir") + ReadMatrix.slash + "\\src\\main\\java\\ir\\alireza\\sadeghi\\shiraz\\ai\\StartingAI\\Aggressive.txt")
    private var neutralInstance =
        ReadMatrix.readIn(System.getProperty("user.dir") + ReadMatrix.slash + "\\src\\main\\java\\ir\\alireza\\sadeghi\\shiraz\\ai\\StartingAI\\Neutral.txt")
    private var defensiveInstance =
        ReadMatrix.readIn(System.getProperty("user.dir") + ReadMatrix.slash + "\\src\\main\\java\\ir\\alireza\\sadeghi\\shiraz\\ai\\StartingAI\\Defensive.txt")

    fun randInstanceCreation() {
        logger.traceEntry()
        ReadMatrix.fileNumber = 0
        val randNeutralInstance = neutralInstance
        val randAggressiveInstance = aggressiveInstance
        val randDefensiveInstance = defensiveInstance
        val randomInterval = 0.1
        for (i in randNeutralInstance!!.indices) {
            for (j in randNeutralInstance[0]!!.indices) {
                randNeutralInstance[i]!![j] =
                    neutralInstance!![i]!![j] - randomInterval + Math.random() * 2 * randomInterval // midpoint of [a,b] is value of NeutralInstance[i][j]
                if (randNeutralInstance[i]!![j] < 0) {
                    randNeutralInstance[i]!![j] = 0.0
                }
            }
        }
        for (i in randAggressiveInstance!!.indices) {
            for (j in randAggressiveInstance[0]!!.indices) {
                randAggressiveInstance[i]!![j] =
                    aggressiveInstance!![i]!![j] - randomInterval + Math.random() * 2 * randomInterval // midpoint of [a,b] is value of AggressiveInstance[i][j]
                if (randAggressiveInstance[i]!![j] < 0) {
                    randAggressiveInstance[i]!![j] = 0.0
                }
            }
        }
        for (i in randDefensiveInstance!!.indices) {
            for (j in randDefensiveInstance[0]!!.indices) {
                randDefensiveInstance[i]!![j] =
                    defensiveInstance!![i]!![j] - randomInterval + Math.random() * 2 * randomInterval // midpoint of [a,b] is value of DefensiveInstance[i][j]
                if (randDefensiveInstance[i]!![j] < 0) {
                    randDefensiveInstance[i]!![j] = 0.0
                }
            }
        }
        normalize(randNeutralInstance as Array<DoubleArray?>?)
        normalize(randAggressiveInstance as Array<DoubleArray?>?)
        normalize(randDefensiveInstance as Array<DoubleArray?>?)
        ReadMatrix.readOut(randNeutralInstance, "Trainers")
        ReadMatrix.readOut(randAggressiveInstance, "Trainers")
        ReadMatrix.readOut(randDefensiveInstance, "Trainers")
    }

    //Normalisation:
    private fun normalize(inputMatrix: Array<DoubleArray?>?) {
        for (l in inputMatrix!!.indices) {
            var sum = 0.0
            for (m in inputMatrix[0]!!.indices) {
                sum += inputMatrix[l]!![m]
            }
            for (n in inputMatrix.indices) {
                inputMatrix[l]!![n] = inputMatrix[l]!![n] / sum
            }
        }
    }
}